pennylane/ops/op_math/exp.py

```diff
@@ -240,11 +266,115 @@ def decomposition(self):
         Returns:
             list[PauliRot]: decomposition of the operator
         """
-        if not self.has_decomposition:
-            raise DecompositionUndefinedError
-        new_coeff = math.real(2j * self.coeff)
-        string_base = qml.pauli.pauli_word_to_string(self.base)
-        return [qml.PauliRot(new_coeff, string_base, wires=self.wires)]
+        return self._recursive_decomposition(self.base, self.coeff)
+
+    def _recursive_decomposition(self, base: Operator, coeff: complex):
+        """Decompose the exponential of ``base`` multiplied by ``coeff``.
+
+        Args:
+            base (Operator): exponentiated operator
+            coeff (complex): coefficient multiplying the exponentiated operator
+
+        Returns:
+            List[Operator]: decomposition
+        """
+        if isinstance(base, Tensor) and len(base.wires) != len(base.obs):
+            raise DecompositionUndefinedError(
+                "Unable to determine if the exponential has a decomposition "
+                "when the base operator is a Tensor object with overlapping wires. "
+                f"Received base {base}."
+            )
+
+        with qml.QueuingManager.stop_recording():
+            # Change base to `Sum`/`Prod`
+            if isinstance(base, Hamiltonian):
+                base = qml.dot(base.coeffs, base.ops)
+            elif isinstance(base, Tensor):
+                base = qml.prod(*base.obs)
+
+        if isinstance(base, SProd):
+            return self._recursive_decomposition(base.base, base.scalar * coeff)
```

Comment:
This and `self._ops_with_generator` seem to be the only instance uses in this method. I'm wondering if we can make these static methods or just extract them from the class completely into "helper" functions.

I don't even think it'd be that bad to just calculate `_ops_with_generator` when we need it instead of caching it on individual instances.

Reply:
We also use `self._pauli_rot_decomposition` and `self._trotter_decomposition`. And in `self._trotter_decomposition` we use `self.num_steps`.

We can definitely make all of them static methods, and pass `num_steps` into `_recursive_decomposition`. But what do we gain from this?

Reply:
Just my general preference to avoid having too much source code in the class itself.